cmd/syncer/main.go

```diff
@@ -16,88 +16,6 @@ limitations under the License.
 
 package main
 
-import (
-	"context"
-	"flag"
-	"fmt"
-	"os"
-	"os/signal"
-	"path"
-	"syscall"
-
-	"k8s.io/apimachinery/pkg/util/sets"
-	"k8s.io/client-go/rest"
-	"k8s.io/client-go/tools/clientcmd"
-	"k8s.io/klog/v2"
-
-	nscontroller "github.com/kcp-dev/kcp/pkg/reconciler/namespace"
-	"github.com/kcp-dev/kcp/pkg/syncer"
-)
-
-const numThreads = 2
-
-var (
-	fromKubeconfig  = flag.String("from_kubeconfig", "", "Kubeconfig file for -from cluster.")
-	fromClusterName = flag.String("from_cluster", "", "Name of the -from logical cluster.")
-	toKubeconfig    = flag.String("to_kubeconfig", "", "Kubeconfig file for -to cluster. If not set, the InCluster configuration will be used.")
-	toContext       = flag.String("to_context", "", "Context to use in the Kubeconfig file for -to cluster, instead of the current context.")
-	pclusterID      = flag.String("cluster", "",
-		fmt.Sprintf("ID of the -to cluster. Resources with this ID set in the '%s' label will be synced.", nscontroller.ClusterLabel))
-)
-
 func main() {
-	flag.Usage = func() {
-		fmt.Fprintf(os.Stderr, "Usage of %s [resourceTypes...]:\n", path.Base(os.Args[0]))
-		flag.PrintDefaults()
-	}
-	flag.Parse()
-	syncedResourceTypes := flag.Args()
-	if len(syncedResourceTypes) == 0 {
-		syncedResourceTypes = []string{"deployments.apps"}
-		klog.Infoln("No resource types provided; using defaults.")
-	}
-	klog.Infof("Syncing the following resource types: %s", syncedResourceTypes)
-
-	// Create a client to dynamically watch "from".
-	if *fromClusterName == "" {
-		klog.Fatal("--from_cluster is required")
-	}
-
-	fromConfig, err := clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-		&clientcmd.ClientConfigLoadingRules{ExplicitPath: *fromKubeconfig}, nil).ClientConfig()
-	if err != nil {
-		klog.Fatal(err)
-	}
-
-	var toConfig *rest.Config
-	if *toKubeconfig != "" {
-		var toOverrides clientcmd.ConfigOverrides
-		if *toContext != "" {
-			toOverrides.CurrentContext = *toContext
-		}
-
-		toConfig, err = clientcmd.NewNonInteractiveDeferredLoadingClientConfig(
-			&clientcmd.ClientConfigLoadingRules{ExplicitPath: *toKubeconfig},
-			&toOverrides).ClientConfig()
-		if err != nil {
-			klog.Fatal(err)
-		}
-	} else {
-		toConfig, err = rest.InClusterConfig()
-	}
-	if err != nil {
-		klog.Fatal(err)
-	}
-
-	ctx, cancel := signal.NotifyContext(context.Background(), syscall.SIGTERM, syscall.SIGILL, syscall.SIGINT)
-	defer cancel()
-
-	klog.Infoln("Starting workers")
-	if err := syncer.StartSyncer(ctx, fromConfig, toConfig, sets.NewString(syncedResourceTypes...), *fromClusterName, *pclusterID, numThreads); err != nil {
-		klog.Fatal(err)
-	}
-
-	<-ctx.Done()
-
-	klog.Infoln("Stopping workers")
+	ExecuteCmd()
```

Comment:
I'd look into following a pattern like you see at https://github.com/kubernetes/kubernetes/blob/88f97283393715c7fa870dd90be0fd0d0faad6bd/cmd/kube-controller-manager/controller-manager.go:

```go
func main() {
	command := app.NewControllerManagerCommand()
	code := cli.Run(command)
	os.Exit(code)
}
```

The reason to put the cobra command in a separate package is so others can import it & use it, and we can test it. The example I shared above puts it in `app`, which would be ok, or you could put it in `cmd` (`cmd/syncer/cmd`).

Reply:
great idea, in fact, I already wanted to. I'm on it

Reply:
done